### PR TITLE
=Fix regression of akka.util.ByteString#slice(...)

### LINKED ISSUE #21237: Akka 2.4.9: ByteString.slice fail with out of bound arguments' values
With 2.4.8 there is no exception occured with any values passed to `slice` method.
But with 2.4.9 we have exceptions in some cases.
```
scala> val sample = ByteString("abcdef")
sample: akka.util.ByteString = ByteString(97, 98, 99, 100, 101, 102)

scala> sample.slice(5, -15)
res1: akka.util.ByteString = ByteString()

scala> sample.slice(-5, -15)
res2: akka.util.ByteString = ByteString()

scala> sample.slice(0, sample.length + 10)
res3: akka.util.ByteString = ByteString(97, 98, 99, 100, 101, 102)

scala> sample.slice(1, sample.length + 10)
java.lang.ArrayIndexOutOfBoundsException: 6
  at akka.util.ByteIterator$ByteArrayIterator.next(ByteIterator.scala:38)
  at akka.util.ByteIterator.foreach(ByteIterator.scala:454)
  ......

scala> sample.slice(-10, 1)
java.lang.ArrayIndexOutOfBoundsException: 6
  at akka.util.ByteIterator$ByteArrayIterator.next(ByteIterator.scala:38)
  at akka.util.ByteIterator.foreach(ByteIterator.scala:454)
  .....
```

### PR DESCRIPTION
**Pull request by monkey-mas**
Fixes #21237

Currently slice(...) method unexpectedly throws an error for some arguments. The issue comes from a corner-case of string _slice_ and we can fix this by correctly handling 6 basic patterns(scenarios) described as follows:

Suppose the two input arguments are;
1) from, until <= 0
We check if `until <= 0` is met and return `ByteString.empty` if this is the case.

2) length <= from, until
We check if `length <= from` is met and return `ByteString.empty` if this is the case.

3) from <= 0 && length <= until
We return the byte string itself as we can _compact_ it.

4) 0 <= from <= until <= length
We _slice_ the byte string starting from `from` with the length of `until - from`

5) from < 0
Based on the condition 1, 2, 3 and 4, we know that there're two cases left, i.e., either `from` or `until` is out of the _valid_ range, as we've checked a) both of them are in the valid range in 4 and b) neither of them is in the valid range (or we don't have to actually slice it where `from == 0` and `until == length`) in 3. If `from` is the one that is out of the valid range then `0 <= until <= length` is satisfied, which means we can return _slice_ from `0` to `until`.

6) length < until
Vice versa, if `until` is the one out of the valid range then `0 <= from <= length` is true, which indicates we can return _slice_ from `from` to `length` with the length of `length - from`.

To satisfy the condition 4, 5 and 6, we need to check if `from <= until` is met. To do this, ByteString1#slice(...) has the corresponding logic.

P.S.
I might misunderstand how the slice methods in `ByteString` and `ByteString1` should work, also this change might deteriorate the performance optimization of `slice`. If this is the case, please let me know. I'll do my best to fix my changes again :)

### CODE REVIEW

**Comment by drewhk** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Thanks a lot for the PR! 
Fixes seem to go in the right direction indeed, but I'll look at them in detail in the morning (also checking what Jenkins didn't like here).

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 1]
```diff
@@ -163,8 +163,8 @@ object ByteString {
       else toByteString1.drop(n)
 
     override def slice(from: Int, until: Int): ByteString =
-      if ((from == 0) && (until == length)) this
```
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: why was optimization removed?

**Comment by johanandren** [comment]
Something around compacting is failing with scalacheck now.

**Comment by ktoso** [comment]
I'll take this on, poke around locally.

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 1 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: good point, this is surely safe I think. 
Digging into other cases

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: removing this actually optimisation breaks scalachecks :)

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 2]
```diff
@@ -253,9 +253,11 @@ object ByteString {
       else ByteString1(bytes, startIndex, Math.min(n, length))
 
     override def slice(from: Int, until: Int): ByteString = {
-      if (from <= 0 && until >= length) this // we can do < / > since we're Compact
+      if (from <= 0 && length <= until) this // we can do < / > since we're Compact
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: remove the comment, it does not apply (my bad)

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 3]
```diff
@@ -253,9 +253,11 @@ object ByteString {
       else ByteString1(bytes, startIndex, Math.min(n, length))
 
     override def slice(from: Int, until: Int): ByteString = {
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: the current impl (previous too) is prone to negative values, let's do a small modification to it:

```scala
      val f = Math.max(from, 0)
      if (f <= 0 && length <= until) this
      else {
        val u = Math.max(until, 0)
        if (u <= f) ByteString1.empty
        else if (0 <= f && u <= bytes.length) ByteString1(bytes, startIndex + f, until - f)
        else if (f < 0) ByteString1(bytes, startIndex, u)
        else /*(bytes.length <= u)*/ ByteString1(bytes, startIndex + f, bytes.length - f)
      }
```

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 4]
```diff
@@ -253,9 +253,11 @@ object ByteString {
...
-      if (from <= 0 && until >= length) this // we can do < / > since we're Compact
+      if (from <= 0 && length <= until) this // we can do < / > since we're Compact
       else if (until <= from) ByteString1.empty
-      else ByteString1(bytes, startIndex + from, until - from)
+      else if (0 <= from && until <= bytes.length) ByteString1(bytes, startIndex + from, until - from)
+      else if (from < 0) ByteString1(bytes, startIndex, until)
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: return `this` instead to save the allocation

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 5]
```diff
@@ -253,9 +253,11 @@ object ByteString {
...
+      else /*(until > bytes.length)*/ ByteString1(bytes, startIndex + from, bytes.length - from)
     }
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Performance has dropped quite a bit with this version:

```
was:
   [info] ByteString_dropSliceTake_Benchmark.bss_large_slice_80_80         thrpt   20  264 963 420.976 ± 4259289.143  ops/s

is: 
   [info] ByteString_dropSliceTake_Benchmark.bss_large_slice_80_80         thrpt   40  145 108 906.085 ± 37355729.647  ops/s
```

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 3 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Meh still optimising the above...

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 5 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: I believe this can be minimised to: 

```scala
    override def slice(from: Int, until: Int): ByteString = 
      if (from <= 0) take(until)
      else if (until <= from) ByteString1.empty
      else if (until <= length) ByteString1(bytes, startIndex + from, until - from)
      else ByteString1(bytes, startIndex + from, bytes.length - from)
```

I'm somewhat tired though, can someone help checking that? (esp. negative values are tricky)

Comment (monkey-mas) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Thanks alot for your help and work! I've got two questions about your code!;

a) Do you assume `from < length` is satisfied when `ByteString1#slice(...)` is called? (I checked `if (length <= from) return empty` in `ByteString#slice(...)` on this [line](https://github.com/akka/akka/pull/21248/files#diff-c8291b63bef2cf0ccb0123d3e15e3833R166) to confirm the condition is met.) or we _CAN'T_ expect the condition?

b) In the third condition
```
else if (until <= length) ByteString1(bytes, startIndex + from, until - from)
```
Did you check `until <= length` using [length](https://github.com/monkey-mas/akka/blob/ea4ff374adf66a4ce51f6ee2b45313549f8f33c2/akka-actor/src/main/scala/akka/util/ByteString.scala#L206), seemingly the length of _sliced_ bytes? or do you mean `bytes.length`, the length of _original_ bytes here?

Based on your answers, I'll take a careful look again! :)

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: You're right, my snippet blows up on `slice(10, 1000)`. I need to look more carefully... 
Trying to minimise if branches, slowdown was noticeable hm, if we'd stay around 250M/s that'd be good - you can check in the akka-bench-jmh benchmarks

Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: What about

```scala
override def slice(from: Int, until: Int): ByteString = 
  if (from <= 0 && length <= until) this
  else if (from < until && from < length && until > 0)
    ByteString1(bytes, startIndex + math.max(0, from), math.min(startIndex + length - from, until))
  else ByteString1.empty
```
?

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala [hunk 6]
```diff
@@ -365,13 +365,30 @@ class ByteStringSpec extends WordSpec with Matchers with Checkers {
       ByteStrings(ByteString1.fromString(""), ByteString1.fromString("ab")).dropRight(Int.MinValue) should ===(ByteString("ab"))
     }
     "slice" in {
-      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(0, 1) should ===(ByteString("a"))
       ByteStrings(ByteString1.fromString(""), ByteString1.fromString("a")).slice(1, 1) should ===(ByteString(""))
+      // Can obtain expected results from 6 basic patterns
+      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(-10, 10) should ===(ByteString("a"))
```
Comment (2beaucoup) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: `slice(-10, -5)` would be another edge case

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 5 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Also crashes on some scalachecks.

I went ahead and checked if we actually can just drop it and do it via `take/drop` turns out perf is good enough:

was (but broken):

```
   [info] ByteString_dropSliceTake_Benchmark.bs_large_slice_80_80          thrpt   20  233 017 314.148 ± 7070246.826  ops/s
   [info] ByteString_dropSliceTake_Benchmark.bs_large_slice_129_129        thrpt   20  275 245 086.247 ± 4969752.048  ops/s
   [info] ByteString_dropSliceTake_Benchmark.bss_large_slice_80_80         thrpt   20  264 963 420.976 ± 4259289.143  ops/s
   [info] ByteString_dropSliceTake_Benchmark.bss_large_slice_129_129       thrpt   20  265 477 577.022 ± 4623974.283  ops/s
```

is (and passes everything):

```
[info] ByteString_dropSliceTake_Benchmark.bs_large_slice_129_129   thrpt   40  213658205.432 ± 5877072.789  ops/s
[info] ByteString_dropSliceTake_Benchmark.bs_large_slice_80_80     thrpt   40  209119190.836 ± 6837219.988  ops/s
[info] ByteString_dropSliceTake_Benchmark.bss_large_slice_129_129  thrpt   40  265664253.164 ± 4064734.763  ops/s
[info] ByteString_dropSliceTake_Benchmark.bss_large_slice_80_80    thrpt   40  260233630.884 ± 5312998.911  ops/s
```

I'm tempted to call that good enough for now as we want to cut a fix-release ASAP actually :~
Unless you have more time on your hands to keep figuring out the optimal way here @monkey-mas.

I also added the following edge case tests so that they always run:

```
      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(1, -2) should ===(ByteString(""))
      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(-10, -100) should ===(ByteString(""))
      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(-100, -10) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(100, 10) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(100, 1000) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(-10, -5) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(-2, -5) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(-2, 1) should ===(ByteString("c"))
```


if you want to hack on it more I suggest you use the following config btw:

```scala
  implicit val betterGeneratorDrivenConfig = PropertyCheckConfig().copy(minSuccessful = 1000) 
// just add it on top of the file
```

it found much more corner cases for me than the default settings.

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala (hunk 6 above)
Comment (ktoso) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: Yeah and note that these do not cover ByteString1.slice, so I suggested a few additional ones.
Technically ScalaCheck does cover them, but in practice these tests are much easier to debug/track.

To find more edge cases it's easy to use the below scalacheck config I pasted

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 5 above)
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: @ktoso Did you find any critical usage of it in Http? Otherwise we should go with take/drop for now. Community is of course welcome to contribute with a better implementation later.

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: I just finished a round of benchmarking, we're OK to go with just `drop().take()` for now and if @monkey-mas has the time to optimise it that would be very nice. Overall HTTP impact is not such that we need to worry about it - difference is within error limits pretty much

Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: good!

**Comment by ktoso** [comment]
I have submitted the simplified fix as https://github.com/akka/akka/pull/21294
It attributes you @monkey-mas for the commit, thanks a lot for your help, and we're looking forward to see if you're able to squeeze max performance out of slice - feel free to ping me if you'd like to bounce ideas or not sure how to run benchmarks etc

**Comment by ktoso** [comment]
Closing this PR, please open a new one with the optimised slice once you have it :)
Thanks again!

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 5 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: What's your box btw? I just measured on a silly macbook this time.

PS: it's better to do `-f2 -i20` than `-f1 -i40` it shows if there's a large variance between jvms

Comment (monkey-mas) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: I used a desktop;
```
OS: Ubuntu 14.0.4 (Linux kernel version 3.13.0-93-generic)
CPU: Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
VM version: JDK 1.8.0_77
```

> it's better to do -f2 -i20 than -f1 -i40

got it thanks :) Seems like I still have almost the same results between `v2.4.9` and my commit.
Not quite sure what actually brought us the difference :(
I'll try previous versions where optimization was not done and see what happens :D
